### DISCUSSION TITLE
Triggerer pod failing after helm chart update

### DISCUSSION THREAD

**Discussion by davidnateberg** [apache/airflow#21068] [Q&A]
Since the helm chart update to `1.4.0` I am seeing the following error in the triggerer pod, specifically the `triggerer` container

```
airflow command error: argument GROUP_OR_COMMAND: triggerer subcommand only works with Python 3.7+, see help above.
usage: airflow [-h] GROUP_OR_COMMAND ...
```

When deploying the `1.3.0` version of the chart, it just worked out of the box, so I'm wondering what might have changed. To add to that, I'm really not sure what the triggerer pod is, I don't see it mentioned in the Apache docs for a k8s deployment, theres no triggerer pod shown in the diagram, just scheduler, webserver, and worker pods.

I have not added any config for the triggerer in the values file, so it would be running with whatever the default config is. I am very new to Airflow and k8s, so please let me know what other information is needed. I could add the values file as well.

**Comment by potiuk** [comment]
This was a bug-fix - you are not supposed to run triggerer on Python 3.6 - it would not work anyway (because of lack of async-io support for Python 3.6).

This is a new feature in Airflow 2.2: https://airflow.apache.org/docs/apache-airflow/stable/concepts/deferring.html

I would recommend you to migrate to Python 3.7 at least becaue Python 3.6 reached end of life in December and Airflow 2.3.0 will not support Python 3.6. so you will have to move to higher python version anyway: https://github.com/apache/airflow/blob/main/README.md#requirements

And yeah. You are right. The graph needs updating.

**Comment by potiuk** [comment]
I opened issue to update the overview https://github.com/apache/airflow/issues/21071

**Comment by davidnateberg** [comment]
Thanks so much for your quick response, I am a little confused by what you mean by migrate to python 3.7, I havent altered any of the default settings for the helm chart so I would expect this is pulling the most recent version of the airflow image, which would use python 3.7 or later no? Is there a version I should be pinning in my values file?

**Reply by davidnateberg** [reply]
i tried tagging the image as `2.2.3-python3.7` in the helm chart and the triggerer is no longer failing, thanks again I really appreciate your speedy responses!

**Reply by potiuk** [reply]
Airlfow Chart as all other charts is as configurable as it can so you are free to modify it. You need to understand a bit about configuration parameters and how it uses images:

The ariflow tag "2.2.3" should in principle means "2.2.3-python3.6" by default and for earlier version it was also 3.6 (we only deprecated 3.6 recently and 3.7 will become default in 2.2.4) - the Deprecation is relatively fresh.

Naming conventions for the tags can be found here: 
https://airflow.apache.org/docs/docker-stack/index.html

```
# Default airflow tag to deploy
defaultAirflowTag: "2.2.3"
# Airflow version (Used to make some decisions based on Airflow Version being deployed)
airflowVersion: "2.2.3"
```

Also if you want to disable triggerer - you can just disable it with configuration:

https://airflow.apache.org/docs/helm-chart/stable/parameters-ref.html#triggerer

**Reply by potiuk** [reply]
Glad I could help!

**Reply by andrew-nascimento** [reply]
This worked for me as well. It's interesting that this seems to have started failing for me yesterday (noticed it this morning) even though I hadn't upgraded the chart to 1.4.0 or changed the image tag.

However, I think the concern (for me anyway) is the default helm chart ships with trigger enabled and using an airflow image version incompatible with that.

**Reply by davidnateberg** [reply]
Agreed that it doesnt seem ideal that by default the newest version of Airflow ships with a component that is not compatible with it.

**Reply by potiuk** [reply]
@jedcunningham @kaxil - I think the concerns of @andrew-nascimento and @davidnateberg are pretty valid. 

And I tihnk it comes from the combination of:

* Helm chart comes with the "triggerer" enabled by default
* also the 2.2.3 images and previous versions had 3.6 as default - only recently (planned for 2.2.4) we switched to 3.7 as default

The fact that @andrew-nascimento mentioned that it only happened recently make me think that MAYBE the 2.2.3  image tag we used to have in DockerHub already was 3.7 before ? 

I had refreshed the images recently when fixing https://github.com/apache/airflow/issues/20911 - and I just double checked it - the 2.2.3 image uses python 3.6 (~as it used to be until we changed early January~) and it was supposed to be 3.7 for 2.2.4 I believe (I cherry-picked that change recently). So I wonder (I might not remember that) but maybe our 2.2.3 image was already Python 3.7 before I refreshed it ? 

Unfortunately DockerHub does not give us TAG history, so I am not 100%  sure what it was before I refreshed the image.

My thinking is that Chrt with Triggerer + 2.2.3 Python 3.6 it is pretty inconsistent and we should fix it. We could actually fix it by re-tagging 2.2.3 image to be Python 3.7. Which is kinda good thing to do due to end of life of Python 3.6 and the likelihood of any problems caused by it is very, very littlle. 

In any way, for the future - maybe we should be more explicit in our chart and use "2.2.3-python3.7" as a tag rather than "2.2.3" ? I think this would be a much more following the "explicit is better than implicit" zen of Python :) 

WDYT?

**Reply by jedcunningham** [reply]
Tags `2.2.1`, `2.2.2`, and `2.2.3`, when they were first released, all used python 3.7. I haven't pulled `2.2.3` recently and mine has 3.7. The chart also jumped from 2.1.4 with 1.2.0 to 2.2.1 with 1.3.0, so there never was a released chart with broken defaults (well, until the latest rebuilds).
 
As for being explicit, I don't have a strong preference either way.

**Reply by potiuk** [reply]
Right let me retag th images then! That was my mistake then!

**Reply by potiuk** [reply]
Sorry for the trouble everyone here, that was a hasty mistake when trying to solve an urgent and serious problem induced by MySQL changing the keys out of a sudden and invalidating the old ones.

**Reply by potiuk** [reply]
Done. 

Sorry for the troubles @davidnateberg @andrew-nascimento. The 2.2.3, 2.2.2, 2.2.1 images are now Python 3.7.

Thanks @jedcunningham for clarifications.

**Reply by andrew-nascimento** [reply]
@potiuk - Not a big deal (I'm not live in production), so it was really an oddity that I noticed. Thanks for the quick fix!

As for the tagging (my 2 cents), I think explicit makes sense or adding a note next the tag that the default version is pythonX.X

eg: 
`# Airflow version (Used to make some decisions based on Airflow Version being deployed) - pythonX.X`
`airflowVersion: "2.2.3"`

**Reply by davidnateberg** [reply]
No worries, same boat as @andrew-nascimento I wasn't in production or anything, I appreciate how responsive everyone in this project is
